### PR TITLE
feat: custom visualization support

### PR DESCRIPTION
**Pull request by Alwinator**
See https://github.com/ActivityWatch/activitywatch/issues/453

### CODE REVIEW

File: aw_server/config.py [hunk 1]
```diff
@@ -6,12 +6,14 @@
 port = "5600"
 storage = "peewee"
 cors_origins = ""
+custom_watcher_visualizations = "{}"
```
Comment (ErikBjare) [inline aw_server/config.py:9]: Avoid JSON in the config. Use the TOML format.

Example:

```toml
[server]
...

[server.custom_visualizations]
aw-watcher-xxx = "/home/user/code/aw-watcher-xxx/static"
```

Maybe call it `custom_static` instead of `custom_visualizations` though? (in case it could generalize)

File: aw_server/custom_watcher_pages.py [hunk 2]
```diff
@@ -0,0 +1,44 @@
...
+        start = iso8601.parse_date(request.json["start"])
+        end = iso8601.parse_date(request.json["end"])
+
+        buckets = current_app.api.get_buckets().keys()
+        custom_page_data = {
+            watcher_name: current_app.api.get_events(get_bucket_name_from_watcher_name(buckets, watcher_name),
+                                                     start=start, end=end)
+            for watcher_name in custom_watcher_static_directories.keys()
+        }
+        return jsonify(custom_page_data)
```
Comment (ErikBjare) [inline aw_server/custom_watcher_pages.py:34]: Why is this needed?

There is already API endpoints to get the events (and to make custom queries), why can't the custom static pages get them itself without relying on this extra endpoint?

File: aw_server/main.py [hunk 3]
```diff
@@ -74,6 +76,11 @@ def parse_settings():
         dest="cors_origins",
         help="CORS origins to allow (as a comma separated list)",
     )
+    parser.add_argument(
+        "--custom-watcher-visualizations",
+        dest="custom_watcher_visualizations",
+        help="The custom watcher visualizations as a JSON string. The JSON contains a dict with key: watcher name and value: static folder path.",
+    )
```
Comment (ErikBjare) [inline aw_server/main.py:83]: Maybe have the user pass a comma-separated list of key-value pairs (delimited by `=`) here instead, such as:

`--custom-static aw-watcher-custom=/home/user/code/aw-watcher-custom/static,...`

JSON is not really CLI-friendly, and should really be avoided in CLI arguments whenever possible.

File: aw_server/custom_watcher_pages.py [hunk 4]
```diff
@@ -0,0 +1,44 @@
...
+def get_bucket_name_from_watcher_name(buckets, watcher_name: str):
+    for bucket in buckets:
+        if bucket.startswith(watcher_name):
+            return bucket
+
+    logging.warning(f"Cannot find bucket for watcher {watcher_name}")
+    return None
+
+
+def get_custom_watcher_blueprint(custom_watcher_static_directories):
```
Comment (ErikBjare) [inline aw_server/custom_watcher_pages.py:16]: Maybe call this simply the `get_custom_static_blueprint`? (and make similar naming changes across the file to simplify and generalize)

File: aw_server/custom_watcher_pages.py [hunk 5]
```diff
@@ -0,0 +1,44 @@
+import logging
```
Comment (ErikBjare) [inline aw_server/custom_watcher_pages.py:1]: Would be nice with a docstring describing basic usage/intent :)

File: aw_server/custom_watcher_pages.py [hunk 6]
```diff
@@ -0,0 +1,44 @@
...
+    logging.warning(f"Cannot find bucket for watcher {watcher_name}")
+    return None
+
+
+def get_custom_watcher_blueprint(custom_watcher_static_directories):
+    custom_watcher_blueprint = Blueprint("custom_watcher", __name__, url_prefix="/watcher")
+
+    @custom_watcher_blueprint.route("api/supported_watchers")
+    def custom_watcher_api_supported_pages():
+        return jsonify(list(custom_watcher_static_directories.keys()))
```
Comment (ErikBjare) [inline aw_server/custom_watcher_pages.py:21]: Instead of having a `supported_watchers` endpoint, make the plain `pages/` endpoint (without a `<string:name>` passed) return the available custom static folders.

**Review by ErikBjare — COMMENTED** [review]
Some comments

File: aw_server/custom_watcher_pages.py (hunk 2 above)
Comment (Alwinator) [reply aw_server/custom_watcher_pages.py:34]: I wanted to reduce the traffic if there are many watchers and to be honest, I was not really sure how to use your custom query language.

File: aw_server/config.py (hunk 1 above)
Comment (Alwinator) [reply aw_server/config.py:9]: Sounds good. I will do that.

File: aw_server/custom_watcher_pages.py (hunk 5 above)
Comment (Alwinator) [reply aw_server/custom_watcher_pages.py:1]: Of course, sorry I forgot about that.

File: aw_server/custom_watcher_pages.py (hunk 6 above)
Comment (Alwinator) [reply aw_server/custom_watcher_pages.py:21]: That makes sense.

File: aw_server/custom_watcher_pages.py (hunk 4 above)
Comment (Alwinator) [reply aw_server/custom_watcher_pages.py:16]: Okay.

File: aw_server/main.py (hunk 3 above)
Comment (Alwinator) [reply aw_server/main.py:83]: Okay,

File: aw_server/main.py [hunk 7]
```diff
@@ -74,6 +75,11 @@ def parse_settings():
         dest="cors_origins",
         help="CORS origins to allow (as a comma separated list)",
     )
+    parser.add_argument(
+        "--custom-static",
+        dest="custom_static",
+        help="The custom static directories. Format: watcher_name=path,wacher_name2=path2,...",
```
Comment (ErikBjare) [inline aw_server/main.py:81]:
```suggestion
        help="The custom static directories. Format: watcher_name=path,watcher_name2=path2,...",
```

File: aw_server/server.py [hunk 8]
```diff
@@ -90,14 +90,18 @@ def _config_cors(cors_origins: List[str], testing: bool):
...
-    host: str,
-    port: int,
-    testing: bool = False,
-    cors_origins: List[str] = [],
+        storage_method,
+        host: str,
+        port: int,
+        testing: bool = False,
+        cors_origins: List[str] = [],
+        custom_static: Dict[str, str] = dict()
```
Comment (ErikBjare) [inline aw_server/server.py:98]: Not sure why indentation changed?

```suggestion
    storage_method,
    host: str,
    port: int,
    testing: bool = False,
    cors_origins: List[str] = [],
    custom_static: Dict[str, str] = dict()
```

File: aw_server/custom_static.py [hunk 9]
```diff
@@ -0,0 +1,69 @@
...
+        start = iso8601.parse_date(request.json["start"])
+        end = iso8601.parse_date(request.json["end"])
+
+        buckets = current_app.api.get_buckets().keys()
+        custom_page_data = {
+            watcher_name: current_app.api.get_events(get_bucket_name_from_watcher_name(buckets, watcher_name),
+                                                     start=start, end=end)
+            for watcher_name in custom_static_directories.keys()
+        }
+        return jsonify(custom_page_data)
```
Comment (ErikBjare) [inline aw_server/custom_static.py:52]: This is my only pet peeve with this PR.

Custom visualizations should really get their data themselves using the API (with aw-client-js for example). There's no need for this special endpoint.

File: aw_server/custom_static.py [hunk 10]
```diff
@@ -0,0 +1,69 @@
+"""
+Contains endpoints, as well as utility functions for custom static content.
```
Comment (ErikBjare) [inline aw_server/custom_static.py:2]: Throw an "experimental" notice here and I'd be willing to merge it :)

```suggestion
Contains endpoints, as well as utility functions for custom static content.

NOTE: Experimental, not (yet) implemented in aw-server-rust.
```

File: aw_server/custom_static.py [hunk 11]
```diff
@@ -0,0 +1,69 @@
...
+
+Idea: Allow custom watchers to extend the Web UI and to write custom visualizations completely independently and free.
+
+Usage for the watcher developer:
+- You can use Vanilla JavaScript, Vue, React, ... as long as you have static content at the end.
+
+- Register your watcher visualization in the config:
+
+[server.custom_static]
+aw-watcher-mywatcher=/home/user/path/to/static/html/dir
```
Comment (ErikBjare) [inline aw_server/custom_static.py:12]:
```suggestion
aw-watcher-example=/home/user/path/to/static_dir/
```

**Review by ErikBjare — COMMENTED** [review]
I would like to merge this. I really like the idea of experimenting with new visualizations through this method.

Looks really good. The only thing is that I don't see the need for the `get_data` endpoint, as it's a bit redundant, so probably just drop that and we'll merge it? :)

**Comment by Alwinator** [comment]
@ErikBjare I am glad that you want to merge the visualization. I implemented the changes according to your feedback in the aw-server. Since the get_data endpoint is gone, I also had to remove it from the aw-webui. There is only one missing: since the data does not get passed to the custom visualization now, the start/end date (that you can pick in the activity watch GUI) needs to be passed to the visualization. I am not sure where to get it from, so I think it would be better if you could implement this tiny last bit. Just replace my null values [here](https://github.com/ActivityWatch/aw-webui/pull/301/files#diff-7bf71b8cbe456c93ccbfc35f7e02130e0801330a8db17dc58821ee9b2b0cb754R20-R22).

Thanks for merging!

**Comment by Alwinator** [comment]
@ErikBjare Yay! Thanks for merging!

**Comment by ErikBjare** [comment]
@Alwinator Thanks for contributing! (and sorry for the delay)

I'm working on an example for the aw-watcher-input watcher to demonstrate how to use the feature, so hopefully that'll be ready sometime next week :)

**Comment by Alwinator** [comment]
@ErikBjare No problem! Sounds great!

**Comment by ErikBjare** [comment]
Here's my super-basic initial example: https://github.com/ActivityWatch/aw-watcher-input/pull/12
